python-spec/src/somacore/query/eager_iter.py

```diff
@@ -0,0 +1,38 @@
+from concurrent import futures
+from typing import Iterator, Optional, TypeVar
+
+_T = TypeVar("_T")
+
+
+class _EagerIterator(Iterator[_T]):
+    def __init__(
+        self,
+        iterator: Iterator[_T],
+        pool: Optional[futures.Executor] = None,
+    ):
+        super().__init__()
+        self.iterator = iterator
+        self._pool = pool or futures.ThreadPoolExecutor()
+        self._own_pool = pool is None
+        self._future: futures.Future[_T] = self._pool.submit(next, self.iterator)  # type: ignore
+
+    def __next__(self) -> _T:
+        try:
+            res: _T = self._future.result()
+            self._future = self._pool.submit(next, self.iterator)  # type: ignore
+            return res
+        except StopIteration:
+            self._cleanup()
+            raise
+
+    def _cleanup(self) -> None:
+        if self._own_pool:
+            self._pool.shutdown()
+
+    def __del__(self) -> None:
```

Comment:
maybe add a helpful comment to the reader why a `__del__` operator is even on the table here -- the reader (including myself) would surely learn something, at just the cost of a sentence or two here